### PR TITLE
Add sliding window attention bias 

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

This PR adds a new attnetion-bias __torch_function__ designed to interact with SDPA. This implements sliding window and updates "aten.sdpa_flash"  to expose the window_size_left and window_size_right arguments. With this kwarg added we can dispatch to the FAv2 impl if the necessary constraints are met.

#### Reviewer Highlights
- One notable difference form FAv2 is that `-1` is used to represent an unbounded window size, while here we accept an optional int, a value of `None` means `window_size_* = inf`. 
- When seqlen_q > seq_len_kv the behavior can be potentially un-intuitive. If you are using the UPPER_LEFT variant, Queries  at indexes > then seq_len_q can have non masked out entries. This is because the window_size_left will allow the query at position(i) to attend to keys at position(i-window_size_left). An analgous argument can be made for LOWER_RIGHT. See this test: https://github.com/pytorch/pytorch/pull/120143/files#diff-14009de3c2bccda8ced7e573adbed54f2a57306659b3bc14e2bfb464f1a3ae17R3523-R3533.
Should we ban this? or just warn?

### Mask defintion

The [Mistral paper](https://arxiv.org/abs/2310.06825) has a very nice visual of the bias:
Screenshot 2024-03-05 at 1 03 30 PM

The sliding window attention bias is parametrized by two values, window_size_left and window_size_right. For any given query `i` this parameters define the "window" or neighborhood for which the query is allowed to attend to; 
```
∀ j ∈ {-window_size_left, -window_size_left + 1, ... 0, ..., window_size_right -1, window_size_right}
attention_bias(Query[i], Key[i + j] = 0

and 
∀ j ∈ [-infinity, window_size_left -1] U [window_size_right + 1, infinity] 
attention-bias(Query[i], Key + j] = -inf
 ```
#### Current Bias API:
In order to construct a mask we require the following args:
```Python
  variant (CausalVariant): The type of causal bias to use (either UPPER_LEFT or LOWER_RIGHT).
  window_size_left (Optional[int]): The size of the left side of the window. If none, the window is unbounded.
  window_size_right (Optional[int]): The size of the right side of the window. If none, the window is unbounded.
  seq_len_q (int): The sequence length of the query tensor.
  seq_len_kv (int): The sequence length of the key/value tensor.
```

#### CausalVariant
The CausalVariant is used to specify the behavior of the attention mechanism when seq_len_q (the length of the query sequence) is not equal to seq_len_kv (the length of the key and value sequences).
There are two possible variants:
- `UPPER_LEFT`: This variant specifies that the first row in the attention matrix corresponds to the query at position/index 0. In other words `bias[0, i]` is the attention bias term between query[0] and key[i]. 
- `LOWER_RIGHT`: This variant specifies that the last row in the attention matrix corresponds to the query at position/index seq_len_kv. In other words  `bias[-1, i]` is the attention bias term between query[seq_len_kv] and key[i].

#### Diagonal attention ( Only attend to tokens at the same position as yourself in the sequence):

```Python
>>> from torch.nn.attention.bias import SlidingWindowBias, CausalVariant
>>> bias = SlidingWindowBias(CausalVariant.UPPER_LEFT, 0, 0, seq_len_q=8, seq_len_kv=8)
>>> bias = SlidingWindowBias(CausalVariant.UPPER_LEFT, 0, 0, seq_len_q=3, seq_len_kv=3)
>>> print(bias._materialize())
tensor([[ True, False, False],
        [False,  True, False],
        [False, False,  True]])
```

#### Causal Attention
``` Python
>>> bias = SlidingWindowBias(CausalVariant.UPPER_LEFT, None, 0, seq_len_q=3, seq_len_kv=3)
>>> print(bias._materialize())
tensor([[ True, False, False],
        [ True,  True, False],
        [ True,  True,  True]])
```

#### Mistral like attention with a backwards context of 2(matches graphic above)
``` Python
>>> bias =bias = SlidingWindowBias(CausalVariant.UPPER_LEFT, 2, 0, seq_len_q=5, seq_len_kv=5)
>>> print(bias._materialize())
tensor([[ True, False, False, False, False],
        [ True,  True, False, False, False],
        [ True,  True,  True, False, False],
        [False,  True,  True,  True, False],
        [False, False,  True,  True,  True]])
```

#### Edge Cases
As you can see from the example above, when seq_len_q == seq_len_kv there is no wrapping since the query and key positions are aligned.

As with the CausalBias mask there exists fully masked out rows with when seq_len_q > seq_len_kv and aligning to lower_right. This matrix doesn't make semantic sense as you have queries at negative index positions.
```Python
>>> bias = SlidingWindowBias(CausalVariant.LOWER_RIGHT, 0, 0, seq_len_q=5, seq_len_kv=2)
>>> print(bias._materialize())
tensor([[False, False],
        [False, False],
        [False, False],
        [ True, False],
        [False,  True]])
```

In fact for this materialization formula we have wonky behavior in this case:
``` Python
bias = SlidingWindowBias(CausalVariant.LOWER_RIGHT, 0, 2, seq_len_q=5, seq_len_kv=2)
>>> print(bias._materialize())
tensor([[False, False],
        [ True, False],
        [ True,  True],
        [ True,  True],
        [False,  True]])
```
the row at bias[-3] should in theory be fully masked but since this would correspond to the query(-1) token.

### Usage with SDPA:
As with the CausalBias this class is expected to only be used with SDPA and does have fused kernel support when the input satisfies all of FAv2s constraints.

Ex:
```Python
from torch.nn.attention.bias import SlidingWindowBias, CausalVariant
from torch.nn.attention import sdpa_kernel, SDPBackend
from torch.nn.functional import scaled_dot_product_attention

bsz=16
num_heads=16
seq_len_q=2048
seq_len_kv=2048
head_dim=64
window_size_left = 1023

# Attend to tokens at position [i - window_size_left, i]
bias = SlidingWindowBias(CausalVariant.UPPER_LEFT, window_size_left, 0, seq_len_q, seq_len_kv)
query = torch.randn(bsz, num_heads, seq_len_q, head_dim, dtype=torch.bfloat16, device="cuda")
key = torch.randn(bsz, num_heads, seq_len_kv, head_dim, dtype=torch.bfloat16, device="cuda")
value = torch.randn(bsz, num_heads, seq_len_kv, head_dim, dtype=torch.bfloat16, device="cuda")

with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
    out_flash = scaled_dot_product_attention(query, key, value, bias)
```

See https://github.com/pytorch/pytorch/issues/119653 for more context

### Performance
```Shell

Results for SLIDING_WINDOW:
+-----------------+-------+------------+-----------+-----------+-----------+-----------+----------------+----------------+------------------+-------------------+----------+
|     Metric      | Value | batch_size | num_heads | q_seq_len | k_seq_len | embed_dim |     dtype      |   bias_type    | window_size_left | window_size_right | head_dim |
+-----------------+-------+------------+-----------+-----------+-----------+-----------+----------------+----------------+------------------+-------------------+----------+
| Average Speedup | 1.40x |            |           |           |           |           |                |                |                  |                   |          |
|   Min Speedup   | 0.94x |     16     |    32     |     1     |   2048    |   2048    | torch.bfloat16 | SLIDING_WINDOW |       4096       |        32         |    64    |
|   Max Speedup   | 1.97x |     1      |    32     |    512    |   4097    |   2048    | torch.bfloat16 | SLIDING_WINDOW |       1024       |         0         |    64    |
+-----------------+-------+------------+-----------+-----------+-----------+-----------+----------------+----------------+------------------+-------------------+----------+
```

### TODO
- [x] Update the meta/Fake implementation with the new kwarg signature: Since the meta funcs have been moved to fake impls that use, args, kwargs: https://github.com/pytorch/pytorch/pull/120682 I dont think any updates need to be made
- [z ] Benchmark Speed ups

cc @mikaylagawarecki

### CODE REVIEW

**Comment by drisspg** [comment]
@desertfire Do you know what the ramifications are of the failing ABI compat test for aotinductor, in theory this should be compatible since I am add kwarg only options to private ATen ops but not sure if there is documentation / detail on the process for doing so

**Comment by desertfire** [comment]
> @desertfire Do you know what the ramifications are of the failing ABI compat test for aotinductor? In theory this should be compatible since I am adding kwarg only options to private ATen ops but not sure if there is documentation / detail on the process for doing so

We already have a v2 `aoti_torch__scaled_dot_product_flash_attention_v2`  for some historical reason, https://github.com/pytorch/pytorch/blob/b9087f857147da37c122cb70955d791565ff1ed6/torch/csrc/inductor/aoti_torch/shim_common.cpp#L376. Normally, you will need to create a similar `v3` version with additional parameters you are adding here, and change the codegen rule here, https://github.com/pytorch/pytorch/blob/b9087f857147da37c122cb70955d791565ff1ed6/torch/_inductor/ir.py#L4954.

But I have a recent PR touching some relevant pieces, https://github.com/pytorch/pytorch/pull/120513. So things are slightly messy here. If you rebase after my PR, it should work for OSS, but there will be issues for internals. If you think this is the only remaining blocker, I can work out a patch PR to unblock.

File: torch/nn/attention/bias.py [hunk 1]
```diff
@@ -351,3 +347,180 @@ def causal_lower_right(*size) -> CausalBias:
...
+                )[0]
+                return _postprocess_flash_output(out, og_head_size)
+            else:
+                _raise_kernel_warnings(sdpa_params)
+                # We cant use flash attention the only support for lower right is via materialization
+                return scaled_dot_product_attention(
+                    query,
+                    key,
+                    value,
+                    attn_mask=attn_mask._materialize(query.device),
```
Comment (cpuhrsch) [inline torch/nn/attention/bias.py:498]: Should we give people a way to hard error when this happens?
